File: apps/web/src/deepLinks.ts
export const DEEP_LINK_VERSION = "1.0" as const;

export type DeepLinkTarget =
  | "home"
  | "job"
  | "similarity"
  | "design_review"
  | "knowledge"
  | "process_trial"
  | "cae"
  | "hmi"
  | "rule_profile"
  | "ingestion_batch";

export type DeepLinkContext = {
  deep_link_version: typeof DEEP_LINK_VERSION;
  target: DeepLinkTarget;
  refs: Record<string, string>;
  correlation_id: string;
};

export type DeepLinkError = {
  code:
    | "DEEP_LINK_VERSION_UNSUPPORTED"
    | "DEEP_LINK_TARGET_INVALID"
    | "DEEP_LINK_REF_INVALID";
  message: string;
};

export type DeepLinkState = { context: DeepLinkContext | null; error: DeepLinkError | null };

const targetRefs: Record<DeepLinkTarget, { required: string[]; optional: string[] }> = {
  home: { required: [], optional: [] },
  job: { required: ["job_id"], optional: [] },
  similarity: { required: ["search_id"], optional: ["candidate_id"] },
  design_review: { required: ["review_id"], optional: ["finding_id"] },
  knowledge: { required: ["knowledge_search_id"], optional: ["citation_id"] },
  process_trial: { required: ["process_search_id"], optional: ["case_id"] },
  cae: { required: ["cae_comparison_id"], optional: ["metric_code"] },
  hmi: { required: ["hmi_extraction_id"], optional: [] },
  rule_profile: { required: ["profile_id"], optional: [] },
  ingestion_batch: { required: ["batch_id"], optional: [] },
};

const uuidRefs = new Set([
  "job_id",
  "search_id",
  "candidate_id",
  "review_id",
  "finding_id",
  "knowledge_search_id",
  "citation_id",
  "process_search_id",
  "case_id",
  "cae_comparison_id",
  "hmi_extraction_id",
  "profile_id",
  "batch_id",
]);

const forbiddenFields = new Set([
  "token",
  "api_key",
  "tunnel_id",
  "workspace_url",
  "return_url",
  "javascript",
  "permission",
]);

const uuidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;

function failure(code: DeepLinkError["code"], message: string): DeepLinkState {
  return { context: null, error: { code, message } };
}

function correlationId(): string {
  return typeof crypto !== "undefined" && "randomUUID" in crypto
    ? crypto.randomUUID()
    : `deep-link-${Date.now()}`;
}

export function parseDeepLink(search: string): DeepLinkState {
  const params = new URLSearchParams(search);
  if ([...params.keys()].length === 0) return { context: null, error: null };
  for (const key of new Set(params.keys())) {
    if (params.getAll(key).length > 1) {
      return failure("DEEP_LINK_REF_INVALID", `Duplicate parameter: ${key}`);
    }
  }

  const version = params.get("deep_link_version");
  if (version !== DEEP_LINK_VERSION) {
    return failure(
      "DEEP_LINK_VERSION_UNSUPPORTED",
      "This Mold AI link uses an unsupported contract version.",
    );
  }
  const rawTarget = params.get("target");
  if (!rawTarget || !(rawTarget in targetRefs)) {
    return failure("DEEP_LINK_TARGET_INVALID", "This Mold AI target is not supported.");
  }
  const target = rawTarget as DeepLinkTarget;
  const schema = targetRefs[target];
  const allowed = new Set(["deep_link_version", "target", ...schema.required, ...schema.optional]);
  for (const key of params.keys()) {
    if (forbiddenFields.has(key) || !allowed.has(key)) {
      return failure("DEEP_LINK_REF_INVALID", `Unexpected parameter: ${key}`);
    }
  }

  const refs: Record<string, string> = {};
  for (const key of [...schema.required, ...schema.optional]) {
    const value = params.get(key) || "";
    if (!value) {
      if (schema.required.includes(key)) {
        return failure("DEEP_LINK_REF_INVALID", `Missing parameter: ${key}`);
      }
      continue;
    }
    if (
      value.length > 128 ||
      /[\u0000-\u001f\u007f]/.test(value) ||
      (uuidRefs.has(key) && !uuidPattern.test(value))
    ) {
      return failure("DEEP_LINK_REF_INVALID", `Invalid parameter: ${key}`);
    }
    refs[key] = value;
  }
  return {
    context: {
      deep_link_version: DEEP_LINK_VERSION,
      target,
      refs,
      correlation_id: correlationId(),
    },
    error: null,
  };
}

export function reportDeepLinkEvent(
  event: "opened" | "resolved" | "failed",
  context: DeepLinkContext,
  detail: { status: string; latency_ms?: number },
): void {
  window.dispatchEvent(
    new CustomEvent(`mold-ai:deep-link-${event}`, {
      detail: {
        contract_version: context.deep_link_version,
        target: context.target,
        correlation_id: context.correlation_id,
        ...detail,
      },
    }),
  );
}

File: services/platform-api/platform_core/deep_links.py
from __future__ import annotations

import uuid
from dataclasses import dataclass
from urllib.parse import urlencode, urlsplit, urlunsplit

DEEP_LINK_VERSION = "1.0"

TARGET_REFS: dict[str, tuple[frozenset[str], frozenset[str]]] = {
    "home": (frozenset(), frozenset()),
    "job": (frozenset({"job_id"}), frozenset()),
    "similarity": (frozenset({"search_id"}), frozenset({"candidate_id"})),
    "design_review": (frozenset({"review_id"}), frozenset({"finding_id"})),
    "knowledge": (frozenset({"knowledge_search_id"}), frozenset({"citation_id"})),
    "process_trial": (frozenset({"process_search_id"}), frozenset({"case_id"})),
    "cae": (frozenset({"cae_comparison_id"}), frozenset({"metric_code"})),
    "hmi": (frozenset({"hmi_extraction_id"}), frozenset()),
    "rule_profile": (frozenset({"profile_id"}), frozenset()),
    "mold_plan": (frozenset({"mold_plan_id"}), frozenset({"resolution_id"})),
    "ingestion_batch": (frozenset({"batch_id"}), frozenset()),
}

UUID_REFS = frozenset(
    {
        "job_id",
        "search_id",
        "candidate_id",
        "review_id",
        "finding_id",
        "knowledge_search_id",
        "citation_id",
        "process_search_id",
        "case_id",
        "cae_comparison_id",
        "hmi_extraction_id",
        "profile_id",
        "mold_plan_id",
        "resolution_id",
        "batch_id",
    }
)

FORBIDDEN_FIELDS = frozenset(
    {
        "token",
        "api_key",
        "tunnel_id",
        "workspace_url",
        "return_url",
        "javascript",
        "permission",
    }
)


class DeepLinkConfigurationError(ValueError):
    pass


class DeepLinkValidationError(ValueError):
    pass


def _canonical_uuid(value: str) -> str:
    try:
        parsed = uuid.UUID(value)
    except (ValueError, AttributeError) as exc:
        raise DeepLinkValidationError("Deep-link identifiers must be canonical UUIDs.") from exc
    canonical = str(parsed)
    if canonical != value:
        raise DeepLinkValidationError("Deep-link identifiers must be canonical UUIDs.")
    return canonical


def normalize_entry_origin(value: str, *, allow_local: bool = False) -> str:
    candidate = value.strip()
    parsed = urlsplit(candidate)
    host = (parsed.hostname or "").lower()
    local_http = allow_local and parsed.scheme == "http" and host == "localhost"
    if parsed.scheme != "https" and not local_http:
        raise DeepLinkConfigurationError("The Web entry URL must use HTTPS.")
    if not host or (
        not local_http
        and (host == "localhost" or host.endswith(".localhost") or host.endswith(".invalid"))
    ):
        raise DeepLinkConfigurationError("The Web entry URL must use a deployable hostname.")
    if parsed.username or parsed.password:
        raise DeepLinkConfigurationError("The Web entry URL must not contain credentials.")
    if parsed.query or parsed.fragment:
        raise DeepLinkConfigurationError(
            "The Web entry URL must not contain query or fragment data."
        )
    if parsed.path not in {"", "/"}:
        raise DeepLinkConfigurationError("The Web entry URL must be an origin without a path.")
    return urlunsplit((parsed.scheme, parsed.netloc, "", "", ""))


@dataclass(frozen=True)
class DeepLinkBuilder:
    entry_base_url: str
    allow_local: bool = False

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "entry_base_url",
            normalize_entry_origin(self.entry_base_url, allow_local=self.allow_local),
        )

    def build(self, target: str, **refs: str | None) -> str:
        if target not in TARGET_REFS:
            raise DeepLinkValidationError("The deep-link target is not supported.")
        if FORBIDDEN_FIELDS.intersection(refs):
            raise DeepLinkValidationError("Sensitive fields are forbidden in deep links.")

        required, optional = TARGET_REFS[target]
        supplied = {key for key, value in refs.items() if value is not None}
        allowed = required | optional
        if not required.issubset(supplied) or not supplied.issubset(allowed):
            raise DeepLinkValidationError(
                "The deep-link references do not match the target schema."
            )

        params: list[tuple[str, str]] = [
            ("deep_link_version", DEEP_LINK_VERSION),
            ("target", target),
        ]
        for key in sorted(supplied):
            raw_value = str(refs[key])
            if key in UUID_REFS:
                raw_value = _canonical_uuid(raw_value)
            elif not raw_value or len(raw_value) > 128 or any(ord(char) < 32 for char in raw_value):
                raise DeepLinkValidationError("The deep-link reference is invalid.")
            params.append((key, raw_value))
        return f"{self.entry_base_url}/open?{urlencode(params)}"


def deep_link_readiness(value: str) -> dict[str, object]:
    try:
        origin = normalize_entry_origin(value)
    except DeepLinkConfigurationError as exc:
        return {
            "ready": False,
            "contract_version": DEEP_LINK_VERSION,
            "entry_origin": None,
            "error": str(exc),
        }
    return {
        "ready": True,
        "contract_version": DEEP_LINK_VERSION,
        "entry_origin": origin,
        "error": None,
    }
